### ISSUE TITLE
Redo can't restore data to smaller partitions showing error: target partition is smaller than original

### ISSUE THREAD

**Issue by Sfinx** [redorescue/redorescue#55]
"Restore data only" process should compare not between the src/dst partitions sizes but between dst partition size and data size

**Comment by agnivo007** [comment]
As far as I remember, not sure though, somehow it's a no-fix for a reason. @zebradots will clarify though...

**Comment by Sfinx** [comment]
It is showstopper - inability restore to smaller partitions make the tool nearly useless. Clonezilla which is based at the same partclone is able to do this.

**Comment by zebradots** [comment]
@Sfinx, thank you for opening this suggestion. To be clear, this is not Redo's intended use case. Redo's interface strives to remain clean and straightforward, without presenting options that could confuse non-technical users. To this end, the application was expressly designed to restore to the same hardware, with a drive capacity equal to or larger than the original. Clonezilla is fine software and, depending on the level of options required, may be a better solution for some users.

However, in order to accommodate this option for advanced users, it is possible that we could expose a checkbox below the restore image details that, for example, could read "Ignore partition size warnings" to let a requested restore operation proceed despite this. This would permit advanced users to restore to smaller partitions, while also making it clear that doing so is unsupported and is could lead to potential data loss (particularly with partitions cloned using `partclone.dd`). Thoughts?

**Comment by Sfinx** [comment]
@zebradots I fully understand the Redo's clean philosophy and do not intend to break it in any way. Just do not understand why when restoring the **src data** to partition the Redo compares the source **partition size** with dst one ? No need for checkboxes here - just compare eggs with eggs but not eggs with apples. I suppose that data size is stored in backup this is why I'm considering this as a bug. Fully agree with you that restoring partition to partition is the most Redo use case. But if you already made the option "Restore data only" then make it right.  Comments ?

**Comment by zebradots** [comment]
Understood, but we have a few legitimate concerns at this time:

1. Partclone's size required to restore a partition will invariably differ slightly from the calculated/actual size; perhaps you already understand the nature of MiB/MB/mb/megabytes etc., a world that can certainly look a lot like the apples and eggs you describe!
2. If Redo demands a little extra to avoid errors, some users may complain about the addition of the margin; what is ideal, a percentage? A fixed size? Does the filesystem affect this choice?
3. The used space of partitions backed up with `partclone.dd` cannot be calculated.

The logic required to handle #1 and #2 are not trivial. There is no fix for #3, and never will be. Any encrypted volume and certain special partitions must be imaged in this way, and adding this feature will do nothing for those use cases.

While we cannot rule out the possibility of silently supporting these operations in the future, the real impediment to implementing this ability lies in providing support for when it fails. Quite frankly, this imposes a sizable support burden on our unpaid team for a very small number of users, particularly with the ever-decreasing cost of storage.

Any reason the checkbox wouldn't suit your needs?

**Comment by Sfinx** [comment]
Then checkbox is ok ;)

**Comment by agnivo007** [comment]
"Ignore partition size warnings (unsupported feature)" Checkbox will work fine for all I guess.

**Comment by zebradots** [comment]
Excellent! This will be slated for addition (PRs welcome).

**Comment by mzky** [comment]
I am also doing the opposite work, the principle is to shrink the disk, then backup;And then you restore it and you expand it。

For XFS of LVM, a feasible approach，
## Note: It is still being debugged. Exceptions may exist。

shrink:
`
#!/bin/bash
rootpath=/rootpath
imgname=$2root$(date '+%Y%m%d%H%M%S').img
dirname=$1
if [ ! -n "$dirname" ]; then
  dirname="Image"
fi
partpath=/mnt/home/partimag/$dirname
mkdir -p $partpath
imgpath=$partpath/$imgname
echo "### shrunk begin ###"
vgchange -ay /dev/centos
if [ $? -ne 0 ]; then
  echo "Error: disk is not lvm!!!"
  exit 1
fi
mount /dev/mapper/centos-root $rootpath
if [ $? -ne 0 ]; then
  echo "Error: disk is unsupported format!!!"
  exit 1
fi
mount -o remount,rw $rootpath
echo "info: dump root to img"
echo "waiting..."
xfsdump -f $imgpath $rootpath -L pcm -M root >/dev/null
if [ $? -ne 0 ]; then
  echo "Error: xfsdump failed !!!"
  exit 1
fi
sha256sum $imgpath >$imgpath.hash
umount $rootpath
lvremove -y /dev/mapper/centos-root
pvresize --setphysicalvolumesize 44G /dev/sda2 -y
resizepart /dev/sda 2 92274688
lvcreate -y -L 30G -n root centos
mkfs.xfs /dev/mapper/centos-root
mount /dev/mapper/centos-root $rootpath
echo "info: restore img to root"
echo "waiting..."
sha256sum -c <(grep $imgpath $imgpath.hash)
if [ $? -ne 0 ]; then
  echo "Error: verify sha256sum failed !!!"
  exit 1
fi
xfsrestore -f $imgpath $rootpath >/dev/null
if [ $? -ne 0 ]; then
  echo "Error: xfsrestore failed !!!"
  exit 1
fi
lsblk
umount $rootpath
echo "### shrunk end ###"
sleep 3
exit 0
`

expand：
`
#!/bin/bash
rootpath=/rootpath
echo "### expand begin ###"
vgchange -ay /dev/centos
if [ $? -ne 0 ]; then
  echo Error: disk is not lvm !!!
  exit 1
fi
growpart /dev/sda 2
mount /dev/mapper/centos-root $rootpath
lvresize -y -l 100%FREE /dev/mapper/centos-root
xfs_growfs $rootpath
lsblk
umount $rootpath
echo "### expand end ###"
sleep 3
exit 0
`

**Comment by 2k9** [comment]
I'd be really looking forward to have a simple checkbox in the GUI for that.
For now this is a showstopper unfortunately.

@mzky is there a way to implement this script so it runs automatically after a restore job?

**Comment by mzky** [comment]
> I'd be really looking forward to have a simple checkbox in the GUI for that.
> For now this is a showstopper unfortunately.
> 
> @mzky is there a way to implement this script so it runs automatically after a restore job?

GUI --> Just use the partitioning tool
script is for startup (GRUP)

**Comment by PizzaProgram** [comment]
I've just tested the new 3.0.2 version to restore a
 0.1 GB + 99GB windows disk to a 120GB SSD.  (Only 30GB data.)

It failed with the error: _"redorescue target drive is smaller than the original"._
Yes, the original drive I backed up was 500GB, but 400 was unallocated from it !! (Empty) Because I've shrink it before backup. 

This makes this whole app unusable.

 _(Yes, yes, at the end: 
- I've created 2 partitions with Gparted
- played half hour with size because Gparted shows MiB and Redo  GB
- used the "Restore data only" process 
- booted into a LiveCD 
- and repaired the boot + + + 
... so spent 2 hours with all of this including searching and finding this ticked.)_


I'm not saying Redo should be able to resize partitions on the fly,
but it ca 30x easier to do these things with Macrium Reflect or Acronis automatically. 
(Even with the free versions.)

# IMHO Redo should be able to restore a 100GB partition to a 120GB SSD with 1 click!

... otherwise the 2014 version of RedoBackup is still better than this one. 
(Faster + old HW compatible + it had FTP search button + fully mouseless operation was available.)

**Comment by PizzaProgram** [comment]
I'm not saying Redo should be able to resize partitions on the fly,
but it ca 30x easier to do these things with Macrium Reflect or Acronis automatically. 
(Even with the free versions.)

# IMHO Redo should be able to restore a 100GB partition to a 120GB SSD with 1 click!

... otherwise the 2014 version of RedoBackup is still better than this one. 
(Faster + old HW compatible + it had FTP search button + fully mouseless operation was available.)

**Comment by zebradots** [comment]
@PizzaProgram: Thank you for your input. With all due respect, it is much easier to request a particular functionality than it is to provide the requisite level of user support to those who checked the box and thought restoring to a smaller partition would work "since there was an option to do it." (See https://github.com/redorescue/redorescue/issues/55#issuecomment-719755957).

We agree that in a perfect world, `partclone` would be able to detect the amount of storage used and somehow resize the partition accordingly, but this is not the case, particularly with encrypted partitions or any filesystem for which `partclone` cannot determine the data allocation bitmap and must revert to `dd`-like imaging.

The solution for an advanced user like yourself is to shrink the partition beforehand, or restore to a virtual device such as a sparse file that can then be resized as needed. You have the knowledge to do these things which most users do not.

A checkbox option that will most certainly corrupt the restored partition and result in further data loss if not properly resized afterward is precisely the type of issue Redo was designed to avoid, not _cause_.

**Comment by JavaJava2Java** [comment]
I agree with PizzaProgram; I don’t understand why Redo has the same glitch as Clonezilla (Rescuezilla? Whateverzilla?). It’s a nice clone of it, like Honda and Hyundai! (Well, at least in the logo.)
The million-dollar question here is: **why can’t they resize partitions 'on the fly'** when cloning, while other similar tools can?
It’s so frustrating to work twice as hard by resizing partitions beforehand and dealing with all the extra hassle.

Don’t get me wrong—I love Clonezilla, Redo, and Whateverzilla when they do the jobs others won’t or help avoid paying a fortune for occasional use. But why design software that’s limited to **'same-size' disks? It doesn’t even make sense.**
I get that it might be hard to program, but spare me the excuses—you’re not cloning physical disks (sectors, etc.), just data. Partitions are data! Other software can do it, so the concept is proven.

And it’s sad because I love and stay loyal to software that’s reliable and versatile, the kind that doesn’t age. (Miss you, Ghost!!)
